libsolidity/formal/CHC.cpp

```diff
@@ -820,20 +849,67 @@ void CHC::visitDeployment(FunctionCall const& _funCall)
 	defineExpr(_funCall, newAddr);
 }
 
+void CHC::internalFunctionCall(
+	FunctionDefinition const* _funDef,
+	Expression const* _calledExpr,
+	FunctionType const* _funType,
+	std::vector<ASTPointer<Expression const>> _arguments,
+	smtutil::Expression _contractAddressValue
+)
+{
+	solAssert(m_currentContract, "");
+	solAssert(_calledExpr, "");
+	solAssert(_funType, "");
+
+	if (_funDef)
+	{
+		if (m_currentFunction && !m_currentFunction->isConstructor())
+			m_callGraph[m_currentFunction].insert(_funDef);
+		else
+			m_callGraph[m_currentContract].insert(_funDef);
+	}
+
+	m_context.addAssertion(predicate(_funDef, _calledExpr, _funType, _arguments, _contractAddressValue));
+
+	solAssert(m_errorDest, "");
+	connectBlocks(
+		m_currentBlock,
+		predicate(*m_errorDest),
+		errorFlag().currentValue() > 0 && currentPathConditions()
+	);
+	m_context.addAssertion(smtutil::Expression::implies(currentPathConditions(), errorFlag().currentValue() == 0));
+	m_context.addAssertion(errorFlag().increaseIndex() == 0);
+}
+
 void CHC::internalFunctionCall(FunctionCall const& _funCall)
 {
 	solAssert(m_currentContract, "");
 
-	auto function = functionCallToDefinition(_funCall, currentScopeContract(), m_currentContract);
-	if (function)
+	auto funDef = functionCallToDefinition(_funCall, currentScopeContract(), m_currentContract);
+	if (funDef)
 	{
 		if (m_currentFunction && !m_currentFunction->isConstructor())
-			m_callGraph[m_currentFunction].insert(function);
+			m_callGraph[m_currentFunction].insert(funDef);
 		else
-			m_callGraph[m_currentContract].insert(function);
+			m_callGraph[m_currentContract].insert(funDef);
 	}
 
-	m_context.addAssertion(predicate(_funCall));
+	Expression const* calledExpr = &_funCall.expression();
+	auto funType = dynamic_cast<FunctionType const*>(calledExpr->annotation().type);
+	std::vector<ASTPointer<Expression const>> arguments = _funCall.sortedArguments();
+
+	auto contractAddressValue = [this](FunctionCall const& _f) {
+		auto [callExpr, callOptions] = functionCallExpression(_f);
+
+		FunctionType const& funType = dynamic_cast<FunctionType const&>(*callExpr->annotation().type);
+		if (funType.kind() == FunctionType::Kind::Internal)
+			return state().thisAddress();
+		if (MemberAccess const* callBase = dynamic_cast<MemberAccess const*>(callExpr))
+			return expr(callBase->expression());
+		solAssert(false, "Unreachable!");
+	};
+
+	m_context.addAssertion(predicate(funDef, calledExpr, funType, arguments, contractAddressValue(_funCall)));
```

Comment:
Can this call the new overload for `internalFunctionCall`?
It looks like there is nontrivial amount of code duplication among the two overloads.

Reply:
done